backend/redis/conn_mgr.c

```diff
@@ -373,52 +632,21 @@ dbBE_Redis_connection_recoverable_t dbBE_Redis_connection_mgr_conn_recover(
             if( addr == dbBE_Redis_server_info_get_master( server ) )
               continue;
 
-            LOG( DBG_ALL, stderr, "Got replica: %s\n", addr );
-
-            dbBE_Redis_connection_t *repl_conn = dbBE_Redis_connection_mgr_newlink( conn_mgr, addr );
-            if( repl_conn == NULL )
-            {
-              LOG( DBG_ALL, stderr, "Replica: %s not yet ready to connect\n", addr );
-              usleep( 250000 );
-              recoverable = DBBE_REDIS_CONNECTION_RECOVERABLE;
-              break;
-            }
-
-            LOG( DBG_ALL, stderr, "Replica: %s connected\n", addr );
-
-            // assign the connection slot range
-            dbBE_Redis_connection_assign_slot_range( repl_conn,
-                                                     first_slot,
-                                                     last_slot );
-
-            dbBE_Redis_connection_mgr_rm( conn_mgr, broke ); // remove the old connection from recv processing
-            dbBE_Redis_connection_destroy( broke ); // delete the old connection
-
-            // if we connected to a replica that's still in replica mode, we end up with MOVED responses all over the place
-            // so tear it down and come back later...
-            switch( dbBE_Redis_connection_mgr_is_master( conn_mgr, repl_conn ) )
+            recoverable = dbBE_Redis_connection_mgr_connect_check_replica( conn_mgr, locator, server,
+                                                                           broke, addr, first_slot, last_slot );
+            switch( recoverable )
             {
-              case 0:
-                LOG( DBG_ALL, stderr, "Cluster fail-over hasn't happen yet. Retrying later.\n");
-                usleep( 250000 );
-                return DBBE_REDIS_CONNECTION_RECOVERABLE;
-              case 1:
-                LOG( DBG_ALL, stderr, "Replica: %s is voted master by Redis now\n", addr );
-
+              case DBBE_REDIS_CONNECTION_RECOVERED:
                 dbBE_Redis_server_info_update_master( server, n ); // make the connection to this replica the new master
-
-                // update locator
-                dbBE_Redis_locator_associate_range_conn_index( locator,
-                                                               first_slot,
-                                                               last_slot,
-                                                               repl_conn->_index );
-
+                ++recovered;
                 break;
-              default:
+              case DBBE_REDIS_CONNECTION_RECOVERABLE:
+                usleep( 250000 );
                 break;
+              case DBBE_REDIS_CONNECTION_UNRECOVERABLE:
```

Comment:
What happens here?

Reply:
in case of an unrecoverable situation, the sender will cancel all pending requests and make them return as failed (at least that's the intended way).